logstash-core/src/main/java/org/logstash/log/StructuredMessageContextSelector.java

```diff
@@ -0,0 +1,33 @@
+package org.logstash.log;
+
+import org.apache.logging.log4j.core.Logger;
+import org.apache.logging.log4j.core.LoggerContext;
+import org.apache.logging.log4j.core.selector.ClassLoaderContextSelector;
+import org.apache.logging.log4j.message.MessageFactory;
+
+import java.net.URI;
+
+public class StructuredMessageContextSelector extends ClassLoaderContextSelector {
+
+    @Override
+    protected LoggerContext createContext(String name, URI configuration) {
+        return new StructuredMessageContext(name, null, configuration);
+    }
+
+    private class StructuredMessageContext extends LoggerContext {
+
+        StructuredMessageContext(String name, Object externContext, URI configuration) {
+            super(name, null, configuration);
+        }
+
+        @Override
+        protected Logger newInstance(final LoggerContext ctx, final String name, MessageFactory messageFactory)
+        {
+            if (null == messageFactory)
+                messageFactory = StructuredMessageFactory.INSTANCE;
+
+            return super.newInstance(ctx, name, messageFactory);
+        }
+
+    }
+}
```

Comment:
If I understand well, the only reason why you create a custom ContextSelector, is to define a custom default `MessageFactory`, right ?

In this case, it exists a specific log4j2 property to customize default `MessageFactory` : `log4j2.messageFactory`  
Using this property, you don't need to create a custom ContextSelector (which aim is to select current log4j context).  

I think it's better to customize only what is necessary, especially concerning log4j2 internals.